### PR TITLE
allow to apply_instance_data to selected instances

### PR DESCRIPTION
**Pull request by anthrotype**
as discussed yesterday

PTAL

### CODE REVIEW

File: Lib/glyphsLib/builder/instances.py [hunk 1]
```diff
@@ -308,30 +310,39 @@ def set_width_class(ufo, designspace, instance):
...
-
-        glyphs_instance = InstanceDescriptorAsGSInstance(instance)
-        # to_ufo_custom_params(self, ufo, data.parent)  # FIXME: (jany) needed?
+    for designspace_instance in designspace.instances:
+        fname = designspace_instance.filename
+        assert fname is not None, ("instance %r missing required filename" %
+                getattr(designspace_instance, "name", designspace_instance))
+        if include_filenames is not None:
+            fname = os.path.normcase(os.path.normpath(fname))
+            if fname not in include_filenames:
```
Comment (belluzj) [inline Lib/glyphsLib/builder/instances.py]: Should you process `include_filenames` with `normpath` and `normcase` as well?

File: tests/builder/instances_test.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    if instance_names is None:
+        # generate all instances
+        builder.process()
+        include_filenames = None
+    else:
+        # generate only selected instances
+        for name in instance_names:
+            builder.readInstance(("stylename", name))
+        # make relative filenames from paths returned by MutatorMath
+        include_filenames = {os.path.normcase(os.path.normpath(
```
Comment (belluzj) [inline tests/builder/instances_test.py]: I think that if you want the callers of `apply_instance_data` to normpath/normcase themselves, you should document that (or have apply_intance_data take care of it)

**Comment by anthrotype** [comment]
ok, maybe it's better we do the normalisation of the  `include_filenames` argument in here, instead of requiring the caller to do it

File: Lib/glyphsLib/builder/instances.py (hunk 1 above)
Comment (anthrotype) [reply Lib/glyphsLib/builder/instances.py]: > Should you process include_filenames with normpath and normcase as well?

done

**Comment by belluzj** [comment]
One last comment: is it ok to break compatibility with older fontmake like this? After this is merged, some people who might want the new version of glyphsLib will have to update fontmake as well?

**Comment by anthrotype** [comment]
hm, I see what you mean. Maybe we could check if the first positional argument of apply_instance_data is an instance of `InstanceData`, and take the `designspace` from there

**Comment by anthrotype** [comment]
Ok, I made the changes backward compatible to make it work with current fontmake (without https://github.com/googlei18n/fontmake/pull/410)

**Pull request merged**